# Answering Business Questions using SQL

## Introduction

### Preliminary

This Notebook is the conclusion of the ***Intermediate SQL for Data Analysis*** from [dataquest.io](dataquest.io). It is a guided project whose aim is to use all the techniques and skills learnt during the course. We will be working here with a database called *Chinook* The latter contains information about a **fictional** digital music shop - kind of like a mini-iTunes store.
The Chinook database contains information about the artists, songs, and albums from the music shop, as well as information on the shop's employees, customers, and the customers purchases. In total there are 11 tables.
Here's a schema diagram for the Chinook database :
![Schema](database_schema.png)


Dpcumentation for the SQLite chinook.db can be found  [here](https://github.com/lerocha/chinook-database).

#### Blockquotes usage
> I am sometimes using blockquotes as this one, meaning that for the rest of the project I am quoting some elements given by dataquest. For the sake of simplicity and clarity, I estimated that they did not need any reformulation and were immediately usable and convenient for me and the reader.

## Overview of the data

In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

In [2]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

 * sqlite:///chinook.db
Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

As shown above and accordingly to the schema, there are 11 tables in our databases. The columns of each table and the links between the tables can be seen on the schema (Se*Introduction* section). To get familiar with our data we are going to print the first ten rows of the tables **employee** and **track**.

In [3]:
%%sql
SELECT *
FROM employee
LIMIT 10;

 * sqlite:///chinook.db
Done.


[(1, 'Adams', 'Andrew', 'General Manager', None, '1962-02-18 00:00:00', '2016-08-14 00:00:00', '11120 Jasper Ave NW', 'Edmonton', 'AB', 'Canada', 'T5K 2N1', '+1 (780) 428-9482', '+1 (780) 428-3457', 'andrew@chinookcorp.com'),
 (2, 'Edwards', 'Nancy', 'Sales Manager', 1, '1958-12-08 00:00:00', '2016-05-01 00:00:00', '825 8 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 2T3', '+1 (403) 262-3443', '+1 (403) 262-3322', 'nancy@chinookcorp.com'),
 (3, 'Peacock', 'Jane', 'Sales Support Agent', 2, '1973-08-29 00:00:00', '2017-04-01 00:00:00', '1111 6 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 5M5', '+1 (403) 262-3443', '+1 (403) 262-6712', 'jane@chinookcorp.com'),
 (4, 'Park', 'Margaret', 'Sales Support Agent', 2, '1947-09-19 00:00:00', '2017-05-03 00:00:00', '683 10 Street SW', 'Calgary', 'AB', 'Canada', 'T2P 5G3', '+1 (403) 263-4423', '+1 (403) 263-4289', 'margaret@chinookcorp.com'),
 (5, 'Johnson', 'Steve', 'Sales Support Agent', 2, '1965-03-03 00:00:00', '2017-10-17 00:00:00', '7727B 41 Ave', 'Calgary', 'AB', 'Canada', 'T3B 1Y7', '1 (780) 836-9987', '1 (780) 836-9543', 'steve@chinookcorp.com'),
 (6, 'Mitchell', 'Michael', 'IT Manager', 1, '1973-07-01 00:00:00', '2016-10-17 00:00:00', '5827 Bowness Road NW', 'Calgary', 'AB', 'Canada', 'T3B 0C5', '+1 (403) 246-9887', '+1 (403) 246-9899', 'michael@chinookcorp.com'),
 (7, 'King', 'Robert', 'IT Staff', 6, '1970-05-29 00:00:00', '2017-01-02 00:00:00', '590 Columbia Boulevard West', 'Lethbridge', 'AB', 'Canada', 'T1K 5N8', '+1 (403) 456-9986', '+1 (403) 456-8485', 'robert@chinookcorp.com'),
 (8, 'Callahan', 'Laura', 'IT Staff', 6, '1968-01-09 00:00:00', '2017-03-04 00:00:00', '923 7 ST NW', 'Lethbridge', 'AB', 'Canada', 'T1H 1Y8', '+1 (403) 467-3351', '+1 (403) 467-8772', 'laura@chinookcorp.com')]

In [4]:
%%sql
SELECT *
FROM track
LIMIT 10;

 * sqlite:///chinook.db
Done.


[(1, 'For Those About To Rock (We Salute You)', 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 343719, 11170334, 0.99),
 (2, 'Balls to the Wall', 2, 2, 1, None, 342562, 5510424, 0.99),
 (3, 'Fast As a Shark', 3, 2, 1, 'F. Baltes, S. Kaufman, U. Dirkscneider & W. Hoffman', 230619, 3990994, 0.99),
 (4, 'Restless and Wild', 3, 2, 1, 'F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. Dirkscneider & W. Hoffman', 252051, 4331779, 0.99),
 (5, 'Princess of the Dawn', 3, 2, 1, 'Deaffy & R.A. Smith-Diesel', 375418, 6290521, 0.99),
 (6, 'Put The Finger On You', 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 205662, 6713451, 0.99),
 (7, "Let's Get It Up", 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 233926, 7636561, 0.99),
 (8, 'Inject The Venom', 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 210834, 6852860, 0.99),
 (9, 'Snowballed', 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 203102, 6599424, 0.99),
 (10, 'Evil Walks', 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 263497, 8611245, 0.99)]

## Selecting Albums to purchase

> The Chinook record store has just signed a deal with a new record label, and you've been tasked with selecting the first three albums that will be added to the store, from a list of four. All four albums are by artists that don't have any tracks in the store right now - we have the artist names, and the genre of music they produce:

| Artist Name          | Genre   |
|----------------------|---------|
| Regal                | Hip Hop |
| Red Tone             | Punk    |
| Meteor and the Girls | Pop     |
| Slim Jim Bites       | Blues   |

> The record label specializes in artists from the USA, and they have given Chinook some money to advertise the new albums in the USA, so we're interested in finding out which genres sell the best in the USA.

We 'll write a query to find out which genres sell the most tracks in the USA,

In [5]:
%%sql

WITH tracks_sold AS
(
    SELECT
        g.name,
        sum(il.quantity) n_tracks_sold
    FROM genre g
    LEFT JOIN track t on g.genre_id = t.genre_id
    INNER JOIN invoice_line il on t.track_id = il.track_id
    GROUP by 1
    ORDER by 2 DESC
)

SELECT
    name,
    n_tracks_sold,
    ROUND(100* CAST(n_tracks_sold as FLOAT) / (SELECT sum(n_tracks_sold)   
                                         FROM tracks_sold),2) percentage
FROM tracks_sold


 * sqlite:///chinook.db
Done.


[('Rock', 2635, 55.39),
 ('Metal', 619, 13.01),
 ('Alternative & Punk', 492, 10.34),
 ('Latin', 167, 3.51),
 ('R&B/Soul', 159, 3.34),
 ('Blues', 124, 2.61),
 ('Jazz', 121, 2.54),
 ('Alternative', 117, 2.46),
 ('Easy Listening', 74, 1.56),
 ('Pop', 63, 1.32),
 ('Electronica/Dance', 55, 1.16),
 ('Classical', 47, 0.99),
 ('Reggae', 35, 0.74),
 ('Hip Hop/Rap', 33, 0.69),
 ('Heavy Metal', 8, 0.17),
 ('Soundtrack', 5, 0.11),
 ('TV Shows', 2, 0.04),
 ('Drama', 1, 0.02)]

According to the above summary, the genre which sell the most are :
- Rock: more than 55% of the total sales
- Metal: 13%
- Alternative & Punk: 10%

The genres we are interested in are according to the 4 four artists we could promote are :
- Hip-Hop : 0.69 of the sales
- Punk: 10% of the sales
- Blues: 2% of the sales
- Pop: 1.32% of the sales

So if we could promote only 3 artists we should do it for:
- Red Tone  
- Meteor and the Girls
- Slim Jim Bites

## Analyzing Employee Sales Performance

In [6]:
%%sql

SELECT *
FROM employee
WHERE title ='Sales Support Agent'

 * sqlite:///chinook.db
Done.


[(3, 'Peacock', 'Jane', 'Sales Support Agent', 2, '1973-08-29 00:00:00', '2017-04-01 00:00:00', '1111 6 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 5M5', '+1 (403) 262-3443', '+1 (403) 262-6712', 'jane@chinookcorp.com'),
 (4, 'Park', 'Margaret', 'Sales Support Agent', 2, '1947-09-19 00:00:00', '2017-05-03 00:00:00', '683 10 Street SW', 'Calgary', 'AB', 'Canada', 'T2P 5G3', '+1 (403) 263-4423', '+1 (403) 263-4289', 'margaret@chinookcorp.com'),
 (5, 'Johnson', 'Steve', 'Sales Support Agent', 2, '1965-03-03 00:00:00', '2017-10-17 00:00:00', '7727B 41 Ave', 'Calgary', 'AB', 'Canada', 'T3B 1Y7', '1 (780) 836-9987', '1 (780) 836-9543', 'steve@chinookcorp.com')]

In [7]:
%%sql
WITH total_customers as
(
    SELECT
    c.customer_id,
    first_name ||' '|| last_name name,
    support_rep_id,
    sum(i.total) total
FROM customer c
LEFT JOIN invoice i on c.customer_id = i.customer_id
GROUP by 1
ORDER by 3
)

SELECT
    first_name ||' '|| last_name employee_name,
    hire_date,
    sum (tc.total) total_sales
FROM employee e
LEFT JOIN totaL_customers tc on e.employee_id = tc.support_rep_id
WHERE e.title = 'Sales Support Agent'
GROUP BY e.employee_id
ORDER BY total_sales
    


 * sqlite:///chinook.db
Done.


[('Steve Johnson', '2017-10-17 00:00:00', 1393.92),
 ('Margaret Park', '2017-05-03 00:00:00', 1584.0),
 ('Jane Peacock', '2017-04-01 00:00:00', 1731.51)]

On the above summary, we can see that among our three Sales Support Agents :
- Jane Peacock is the best with 1731 \$ sold
- Margaret Park is the 2nd with 1584 \$ sold
- Steve Johnson is the 3rd with 1392 \$ sold

But nevertheless this ranking has to be considered with the hiring date of each of the employees. Indeed Jane was hired (April 2017) one month before Margaret (May 2017) and more than 6 months before Steve (Mid-October 2017). So actually this ranking is not very representative for the employees performances.

## Analyzing Sales by country

Here is below a first query to get a summary for each country **individually**

In [8]:

%%sql

SELECT
    country,
    nb_customers,
    sales_value,
    sales_value / nb_customers avg_sales_customers,
    avg_order
FROM
(
    SELECT
        country,
        count(distinct i.customer_id) nb_customers,
        sum(i.total) sales_value,
        avg(i.total) avg_order
    FROM customer c
    LEFT JOIN invoice i on c.customer_id = i.customer_id
    GROUP BY 1
    ORDER BY 3 DESC
)

 * sqlite:///chinook.db
Done.


[('USA', 13, 1040.4900000000005, 80.03769230769234, 7.942671755725194),
 ('Canada', 8, 535.59, 66.94875, 7.047236842105264),
 ('Brazil', 5, 427.6799999999999, 85.53599999999997, 7.011147540983605),
 ('France', 5, 389.07000000000005, 77.81400000000001, 7.781400000000001),
 ('Germany', 4, 334.61999999999995, 83.65499999999999, 8.161463414634145),
 ('Czech Republic', 2, 273.23999999999995, 136.61999999999998, 9.107999999999999),
 ('United Kingdom', 3, 245.52, 81.84, 8.768571428571429),
 ('Portugal', 2, 185.13, 92.565, 6.3837931034482756),
 ('India', 2, 183.14999999999995, 91.57499999999997, 8.72142857142857),
 ('Ireland', 1, 114.83999999999997, 114.83999999999997, 8.833846153846151),
 ('Spain', 1, 98.01, 98.01, 8.91),
 ('Chile', 1, 97.02000000000001, 97.02000000000001, 7.463076923076924),
 ('Australia', 1, 81.18, 81.18, 8.118),
 ('Finland', 1, 79.2, 79.2, 7.2),
 ('Hungary', 1, 78.21, 78.21, 7.821),
 ('Poland', 1, 76.22999999999999, 76.22999999999999, 7.622999999999999),
 ('Sweden', 1, 75.24, 75.24, 7.523999999999999),
 ('Norway', 1, 72.27000000000001, 72.27000000000001, 8.030000000000001),
 ('Austria', 1, 69.3, 69.3, 7.699999999999999),
 ('Netherlands', 1, 65.34, 65.34, 6.534000000000001),
 ('Belgium', 1, 60.38999999999999, 60.38999999999999, 8.627142857142855),
 ('Italy', 1, 50.49, 50.49, 5.61),
 ('Argentina', 1, 39.6, 39.6, 7.92),
 ('Denmark', 1, 37.61999999999999, 37.61999999999999, 3.761999999999999)]

In [9]:
%%sql

WITH sales_country as
(
    SELECT
        (CASE 
             WHEN nb_customers = 1 THEN 'Other'
             ELSE country 
        END) AS country,
        sum(nb_customers) nb_customers,
        sum(sales_value) sales_value,
        sum(sales_value) / sum(nb_customers) avg_sales_customers,
        avg_order
    FROM
    (
        SELECT
            country,
            count(distinct i.customer_id) nb_customers,
            sum(i.total) sales_value,
            sum(i.total) / count(i.invoice_id) avg_order
        FROM customer c
        LEFT JOIN invoice i on c.customer_id = i.customer_id
        GROUP BY 1
        ORDER BY 3 DESC
    )

    GROUP BY 1
    ORDER BY 3 DESC
),

intermediary as 
(
    SELECT sales_country.*,
            (CASE
                 WHEN country='Other' THEN 1
                 ELSE 0
            END) as sort
    FROM sales_country
)

SELECT
    country,
    nb_customers,
    sales_value,
    avg_sales_customers,
    avg_order
FROM intermediary
ORDER BY sort ASC 
    




 * sqlite:///chinook.db
Done.


[('USA', 13, 1040.4900000000005, 80.03769230769234, 7.942671755725194),
 ('Canada', 8, 535.59, 66.94875, 7.047236842105264),
 ('Brazil', 5, 427.6799999999999, 85.53599999999997, 7.011147540983605),
 ('France', 5, 389.07000000000005, 77.81400000000001, 7.781400000000001),
 ('Germany', 4, 334.61999999999995, 83.65499999999999, 8.161463414634145),
 ('Czech Republic', 2, 273.23999999999995, 136.61999999999998, 9.107999999999999),
 ('United Kingdom', 3, 245.52, 81.84, 8.768571428571429),
 ('Portugal', 2, 185.13, 92.565, 6.3837931034482756),
 ('India', 2, 183.14999999999995, 91.57499999999997, 8.72142857142857),
 ('Other', 15, 1094.9399999999998, 72.996, 3.761999999999999)]

Here is above a summary of the sales by country. It includes : 
- the number of customers
- the total sales value
- the average value of sales per customer
- the average order

Note that all country having only one customer have been gathered and grouped under the "Other" alias.

##  Albums vs Individual Tracks

> The Chinook store is setup in a way that allows customer to make purchases in one of the two ways:
- purchase a whole album
- purchase a collection of one or more individual tracks.

>The store does not let customers purchase a whole album, and then add individual tracks to that same purchase (unless they do that by choosing each track manually). When customers purchase albums they are charged the same price as if they had purchased each of those tracks separately.

>Management are currently considering changing their purchasing strategy to save money. The strategy they are considering is to purchase only the most popular tracks from each album from record companies, instead of purchasing every track from an album.

>We have been asked to find out what percentage of purchases are individual tracks vs whole albums, so that management can use this data to understand the effect this decision might have on overall revenue.

>In order to answer the question, we're going to have to identify whether each invoice has all the tracks from an album. We can do this by getting the list of tracks from an invoice and comparing it to the list of tracks from an album. We can find the album to compare the purchase to by looking up the album that one of the purchased tracks belongs to. It doesn't matter which track we pick, since if it's an album purchase, that album will be the same for all tracks.

This part of code has been copied and pasted from [juliechipko](https://github.com/dataquestio/solutions/blob/master/Mission191Solutions.ipynb ) 

In [10]:
%%sql

WITH invoice_first_track AS
    (
     SELECT
         il.invoice_id invoice_id,
         MIN(il.track_id) first_track_id
     FROM invoice_line il
     GROUP BY 1
    )

SELECT
    album_purchase,
    COUNT(invoice_id) number_of_invoices,
    CAST(count(invoice_id) AS FLOAT) / (
                                         SELECT COUNT(*) FROM invoice
                                      ) percent
FROM
    (
    SELECT
        ifs.*,
        CASE
            WHEN
                 (
                  SELECT t.track_id FROM track t
                  WHERE t.album_id = (
                                      SELECT t2.album_id FROM track t2
                                      WHERE t2.track_id = ifs.first_track_id
                                     ) 

                  EXCEPT 

                  SELECT il2.track_id FROM invoice_line il2
                  WHERE il2.invoice_id = ifs.invoice_id
                 ) IS NULL
             AND
                 (
                  SELECT il2.track_id FROM invoice_line il2
                  WHERE il2.invoice_id = ifs.invoice_id

                  EXCEPT 

                  SELECT t.track_id FROM track t
                  WHERE t.album_id = (
                                      SELECT t2.album_id FROM track t2
                                      WHERE t2.track_id = ifs.first_track_id
                                     ) 
                 ) IS NULL
             THEN "yes"
             ELSE "no"
         END AS "album_purchase"
     FROM invoice_first_track ifs
    )
GROUP BY album_purchase;

 * sqlite:///chinook.db
Done.


[('no', 500, 0.8143322475570033), ('yes', 114, 0.18566775244299674)]

**Conclusion** : most of the sales (more than 80%) are selected unique tracks and not full album purchases. So Chinook sould not continue to buy full albums to records store but rather buy the most popular tracks as they had considered it.